Sources/Algorithms/CommonPrefix.swift

```diff
@@ -0,0 +1,350 @@
+    }
+  }
+  
+  @inlinable
+  internal func normalizeIndex(base: Base.Index, other: Other.Index) -> Index {
+    if base != self.base.endIndex
+        && other != self.other.endIndex
+        && areEquivalent(self.base[base], self.other[other])
+    {
+      return Index(base: base, other: other)
+    } else {
+      return endIndex
+    }
+  }
+  
+  @inlinable
+  public var startIndex: Index {
+    normalizeIndex(base: base.startIndex, other: other.startIndex)
+  }
+  
+  @inlinable
+  public var endIndex: Index {
+    Index(base: base.endIndex, other: other.endIndex)
+  }
+  
+  @inlinable
+  public func index(after index: Index) -> Index {
+    normalizeIndex(
+      base: base.index(after: index.base),
+      other: other.index(after: index.other))
+  }
+  
+  @inlinable
+  public subscript(index: Index) -> Base.Element {
+    base[index.base]
+  }
+}
+
+extension CommonPrefix.Index: Comparable {
+  @inlinable
+  public static func == (lhs: Self, rhs: Self) -> Bool {
+    lhs.base == rhs.base
+  }
+  
+  @inlinable
+  public static func < (lhs: Self, rhs: Self) -> Bool {
+    lhs.base < rhs.base
+  }
+}
+
+extension CommonPrefix.Index: Hashable where Base.Index: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(base)
+  }
+}
+
+extension CommonPrefix: LazySequenceProtocol where Base: LazySequenceProtocol {}
+extension CommonPrefix: LazyCollectionProtocol
+  where Base: LazyCollectionProtocol, Other: Collection {}
+
+//===----------------------------------------------------------------------===//
+// Sequence.commonPrefix(with:)
+//===----------------------------------------------------------------------===//
+
+extension Sequence {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other,
+    by areEquivalent: (Element, Other.Element) throws -> Bool
+  ) rethrows -> [Element] {
+    var iterator = makeIterator()
+    var otherIterator = other.makeIterator()
+    var result: [Element] = []
+    
+    while let next = iterator.next(),
+          let otherNext = otherIterator.next(),
+          try areEquivalent(next, otherNext)
+    {
+      result.append(next)
+    }
+    
+    return result
+  }
+}
+
+extension Sequence where Element: Equatable {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other
+  ) -> CommonPrefix<Self, Other> where Other.Element == Element {
+    CommonPrefix(base: self, other: other, areEquivalent: ==)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// LazySequenceProtocol.commonPrefix(with:)
+//===----------------------------------------------------------------------===//
+
+extension LazySequenceProtocol {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other,
+    by areEquivalent: @escaping (Element, Other.Element) -> Bool
+  ) -> CommonPrefix<Self, Other> {
+    CommonPrefix(base: self, other: other, areEquivalent: areEquivalent)
+  }
```

Comment:
Shouldn't these be using `Elements` and `elements` instead of `Self` and `self`?

Reply:
That's a way to do it, but this is intentional. If you wrap `elements` instead of `self` then the type no longer contains any sign of laziness, so you'll need to wrap it in `LazySequence<...>` again to add the laziness back. But by wrapping `self`, all we need is for `CommonPrefix` to conditionally conform to `LazySequenceProtocol` when the (first) base collection does, in order to propagate the laziness.

`CommonPrefix` also can't unconditionally conform to `LazySequenceProtocol` because it's also returned by the `Sequence` overloads, so in a non-lazy context.